pico: add uart-sample program

KantaTamura opened the pull request:
uart sample program
- `console-hello`
"hello world"をconsoleに出力する
- `console-echo`
consoleからの入力をそのまま送り返す

### build
UART0を使用する
```
$ west build -b rpi_pico --build-dir <***> scsat1-rpi/pico/sample/<console-echo/console-hello>
```

UART1を使用する
```
$ west build -b rpi_pico --build-dir <***> scsat1-rpi/pico/sample/<console-echo/console-hello> -- -DUSE_UART1=1
```

USBを使用する
```
$ west build -b rpi_pico --build-dir <***> scsat1-rpi/pico/sample/<console-echo/console-hello> -- -DUSE_USB=1
```

デバッグメッセージを表示する
```
$ west build -b rpi_pico --build-dir <***> scsat1-rpi/pico/sample/<console-echo/console-hello> -- -DALLOW_DEBUG=1
```

> [!CAUTION]
> 変数を利用するとcmakeのキャッシュに貯まるようなので，一度変数を有効にすると次回以降は `-DUSE_USB=0` のように使用しないことを明示する必要がある

KantaTamura committed:
pico: sample: add output "hello world" to console

console-helloは，"zephyr_console"を切り替えながら
任意のuartデバイスに"hello world"を出力するサンプルプログラム

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura committed:
pico: sample-console-hello: add usb-console

USBのUARTに向けて出力できるように変更

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura committed:
pico: sample: add echo server

`console-hello`と同様にconsoleを切り替えながら
uartのecho-serverとして機能するサンプルプログラム

ref.
- zephyr, UART echo sample

Signed-off-by: KantaTamura <tkanta496@gmail.com>

sasataku commented on pico/sample/console-echo/src/main.c:
```diff
@@ -0,0 +1,127 @@
+    if (!device_is_ready(uart_dev)) {
+		LOG_ERR("UART device not found!");
+		return 0;
+	}
```
インデントがおかしい。他にもいくつかあるので、clang-formatを使ってreformatしてみてください。
https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format

KantaTamura replied:
ほんとですね...vscodeさんの表示だと上手く言ってるのですが...

yashi replied:
あー、 [.editorconfig](https://editorconfig.org/) 置こうか？

sasataku commented on pico/sample/console-hello/src/main.c:
```diff
@@ -23,12 +26,16 @@ int main(void)
+    LOG_INF("application started.");
+
     while (1) {
+        LOG_INF("Log Hellow!");
```
typo.
`Hellow`

sasataku commented on pico/sample/console-echo/src/main.c:
```diff
@@ -0,0 +1,127 @@
+{
+    int ret;
+    if ((ret = uart_irq_update(uart_dev)) != 0) {
+        LOG_ERR("can't start interrupt : %d\n", ret);
```
LOG_XXには、明示的な改行コードは不要。

sasataku commented on pico/sample/console-echo/src/main.c:
```diff
@@ -0,0 +1,127 @@
+
+/* message configuration */
+#define MSG_SIZE 32
+K_MSGQ_DEFINE(uart_msgq, MSG_SIZE, 10, 4); /* message queue (up to 10 messages) */
```
パラメータはKconigにもっていったら？とおもったけど、`zephyr/samples/drivers/uart/echo_bot`をベースにしているからだと気がついた。とりあえずコミットログにでもベースにしたコードを書いておいてほしい。

yashi commented:
`console-hello` って言うより、 `uart-hello`?

yashi commented on pico/sample/console-echo/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+
+set(BOARD, rpi_pico)
```
`,` は使わない。これ CMake 使い始めのとき、必ずやっちゃうやつw

これだと `BOARD,` に `rpi_pico` をセットしているので `BOARD` はセットされてないことになる。なのでビルド時に `-b rpi_pico` を指定しているのでは？

- `west build -b rpi_pico ...` とするなら、 `set(BOaRD rpi_pico)` は不要
- `west build` とするなら、 `,` を外す

KantaTamura replied:
何故かボードが指定されていないと思いました...笑

yashi replied:
`pico/CMakeLists.txt` も別コミット or 別 PR で直してもらえる？前回は気づかなかったみたい

yashi commented on pico/sample/console-echo/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
+if (USE_USB)
+    list(APPEND EXTRA_DTC_OVERLAY_FILE "usb.overlay")
+    list(APPEND EXTRA_CONF_FILE "usb.conf")
```
Kconfig は、 device が active の時に `default y`にしておけばよいよ。 `depends on DT_HAS_` で。
たとえば

zephyr/drivers/i2c/Kconfig.ambiq:
```Kconfig
config I2C_AMBIQ
        bool "AMBIQ I2C driver"
        default y
        depends on DT_HAS_AMBIQ_I2C_ENABLED
        select AMBIQ_HAL
        select AMBIQ_HAL_USE_I2C
        help
          Enable driver for Ambiq I2C.
```

yashi replied:
> Kconfig は、 device が active の時に `default y`にしておけばよいよ。 

あ、違うか。 Kconfig 書きたいんじゃなくて、 `.conf` を書きたいのか

yashi commented on pico/sample/console-echo/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+
+set(BOARD, rpi_pico)
+
+if (USE_UART1)
```
`USE_UART1` と `USE_USB` があると、両方選択できる。それでも後者が選択されるので問題ないが、あえて `-D` で選択できるようにしているんだから

```cmake
cmake_minimum_required(VERSION 3.20)
project(string_comp)

if(CONSOLE STREQUAL uart1)
  message(STATUS one!)
elseif(CONSOLE STREQUAL usb)
  message(STATUS usb!)
else()
  message(STATUS default!)
endif()
```

```shell
cmake . -DCONSOLE=uart1
-- one!
```

```shell
cmake . -DCONSOLE=usb
-- usb!
```

```shell
cmake .
-- default!
```

とか

yashi commented on pico/sample/console-echo/src/main.c:
```diff
@@ -0,0 +1,127 @@
+void serial_callback_handler(const struct device *dev, void *user_data)
+{
+    int ret;
+    if ((ret = uart_irq_update(uart_dev)) != 0) {
```
```c
int ret;

ret = uart_irq_update(uart_dev);
if (ret != 0) {
   ;
}
```
としてくれる？ `if()` の中で関数呼ぶの嫌いなんだよね。関数を呼ぶ方が大事なのに分岐の方が目立つから。で目立つくせに、比較対象が関数よりも右に来るのが嫌。ずっと趣味の問題だと思っていたら、 [The Programmer's Brain](https://www.manning.com/books/the-programmers-brain) にも似たようなこと書いてあった。
あと、 debugger で step するときにも分かれていてほしい。

yashi replied:
とか言いながら、Ruby とかだと、 method なのか property access なのかわからので、

```ruby
if obj.prop != 0 then
    ....
end
```
は、気にならないというw。これは、多分短いからだと思う。長くなったら、Ruby でも嫌い。

KantaTamura replied:
いままで `if()` 内で呼ぶことが多かったのですが，
デバッグのときにやりにくいのはその通りですね...

yashi commented on pico/sample/console-echo/src/main.c:
```diff
@@ -0,0 +1,127 @@
+int main(void)
+{
+/* enable only has DTR line e.g. USB */
+#if IS_ENABLED(DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console), zephyr_cdc_acm_uart))
```
```c
int main() {
    if (DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console), zephyr_cdc_acm_uart)) {
       ...
    }
```
じゃ、だめなんだっけ？

KantaTamura replied:
USBを指定しない限り意味のない条件分岐なのでコンパイル時に弾けたほうが良いかな？
と思いました

yashi replied:
なるほど。 でも

1. `if (0) {...}` のようにコンパイル時に不採用(not taken) が確定する分岐の場合、最近のコンパイラーは分岐をまるっと削除してくれる。 [dead code elimination](https://en.wikipedia.org/wiki/Dead-code_elimination) という optimize が GCC や Clang には入っている。ので Runtime overhead は無い。
2. Preprocessor を使うと、必ずプリプロセスをしないとコンパイルする・しないが分かれるコードになる。場合分けが増えると指数関数的にテストパターンが増える。 `if`文だとコンパイラーは構文チェックを必ずするので、安心。あまり増えすぎると (当たり前だが) コードが大きくなるのでコンパイル時間が増える

KantaTamura replied:
最近のコンパイラは優秀ですね...

ということは `#if` とかは最近だとあまり使うべきではないってことですか？
`#ifdef` とかは用途がありそうですが...

yashi replied:
https://google.com/search?q=ifdef+considered+harmful

必要な場合もあるよ？ 低いレイヤーで OS とか CPU Architecture が違う場合とか。でもアプリの場合はもっと抽象度を上げても良いと思う場合がほとんどだけど。

yashi replied:
たとえばこの場所も `#ifdef` や `IS_ENABLED` じゃなくて `uart_config_get()` で flow control がサポートされているか？ でも良かったり？

yashi commented:
Twsiter と `sample.yaml` によるビルドってできないかな？

- 例) `zephyr/samples/bluetooth/hci_uart/sample.yaml`
- https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html
- https://docs.zephyrproject.org/latest/develop/test/twister.html#twister-script

yashi commented:
> Caution
> 
> 変数を利用するとcmakeのキャッシュに貯まるようなので，一度変数を有効にすると次回以降は `-DUSE_USB=0` のように使用しないことを明示する必要がある

または、 [`west build -p`](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#pristine-builds) を使うとか

KantaTamura committed:
pico: sample-console-hello: change cmake commandline variable

before:
```
cmake . -DUSE_UART1=1 -DUSE_USB=1
```
with the above command,
the USB console is enabled, and the UART1 is ignored.
This is not intuitive, so I will change it
to be specified by the following string.

after:
```
cmake . -DCONSOLE=uart1
cmake . -DCONSOLE=usb
```

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura commented:
> Twsiter と sample.yaml によるビルドってできないかな？

しらべてみます！

yashi commented:
> > Twsiter と sample.yaml によるビルドってできないかな？
> 
> しらべてみます！

時間かかりそうなら、先にこの PRをマージしよう

KantaTamura commented:
> 時間かかりそうなら、先にこの PRをマージしよう

おそらく，本格的に作業できるのは2/7以降だと思うので別のPRにしたほうが良いかな？と思います！

sasataku reviewed (approved):
LGTM

yashi commented on pico/sample/console-hello/src/main.c:
```diff
@@ -0,0 +1,46 @@
+    const struct device *dev = CONSOLE_DEVICE;
+
+    /* enable only has DTR line */
+    if (DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console), zephyr_cdc_acm_uart))
```
`{` が無い。 build error になる

yashi commented on pico/sample/console-hello/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+
+set(BOARD, rpi_pico)
```
`,` がまだある

yashi reviewed (changes requested):
とりあえず、build できるようにしようか

yashi commented:
あとこの repository に入っている `.clang-format` を使ってください。

yashi commented:
この修正したときに、 I2C が正常に動作できることって、確認できていますか？

KantaTamura commented:
> とりあえず、build できるようにしようか

> この修正したときに、 I2C が正常に動作できることって、確認できていますか？

すみません...全部ビルドして実行もできることを確認しました！

KantaTamura commented:
> あとこの repository に入っている .clang-format を使ってください。

vscodeだと，working directoryのトップに配置されてないと認識してくれなかったみたいです

yashi commented on pico/sample/console-echo/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+mainmenu "pico-sample: console-hello"
+
+source "Kconfig.zephyr"
```
これは、あえてここにしているの？もしそうなら、その理由を聞きたい

KantaTamura replied:
僕がKconfigを触るのが初めてなので，いろいろ手探り+勘違いしている可能性が高そうです...すみません...

その部分は，logginのための
```
source "subsys/logging/Kconfig.template.log_config"
```
を追加するために，[ドキュメント](https://docs.zephyrproject.org/latest/develop/application/index.html#application-cmakelists-txt)を見ながら，`Kconfig` を書いたはずです

サンプルの
```
# SPDX-License-Identifier: Apache-2.0

mainmenu "Your Application Name"

# Your application configuration options go here

# Sources Kconfig.zephyr in the Zephyr root directory.
#
# Note: All 'source' statements work relative to the Zephyr root directory (due
# to the $srctree environment variable being set to $ZEPHYR_BASE). If you want
# to 'source' relative to the current Kconfig file instead, use 'rsource' (or a
# path relative to the Zephyr root).
source "Kconfig.zephyr"
```
に追加する形にしてました

yashi commented on pico/sample/console-echo/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+source "Kconfig.zephyr"
+
+module = MAIN
+source "subsys/logging/Kconfig.template.log_config" # for logging
```
`module-str` がないと、エラーにならない？

KantaTamura replied:
[参考にしたドキュメント](https://docs.zephyrproject.org/latest/services/logging/index.html#logging-in-a-module)の
```
module = FOO
module-str = foo
source "subsys/logging/Kconfig.template.log_config"
```
には `module-str` があるので，コピー忘れたように思います．

なくてもビルド自体はできたので気づかなかったです...

yashi reviewed (changes requested):
`west build -t menuconfig` を試してみて。

`app/Kconfig` (今回場合は `console-echo/Kconfig` とか) を用意すると言うことは、 menuconfig などの config 画面を用意すると言う事。 `Kconfig.zephyr` を `source` するのは、 Zephyr のデフォルトの config をロードするということ。なのでどのようながメニューを準備して、どこに Zephyr の menu 画面を出すのかということですよ。

https://docs.zephyrproject.org/latest/build/kconfig/menuconfig.html

たとえば

```kconfig
mainmenu "pico-sample: console-hello"

menu "My Application Config"

module = MAIN
moduel-str = "Hoge fuga"
source "subsys/logging/Kconfig.template.log_config" # for logging

endmenu

source "Kconfig.zephyr"
```
とすると

![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/3049cd33-f2c2-444c-91ab-e1438b0b3312)

こうなる

yashi commented:
ちなみに、いまのままだと、こうなる。一番最後の `(3) <MAIN_LOG_LEVEL>` 。

![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/e8dc7b4a-1623-4db9-a454-cb32ce51962c)

赤字の部分はふつうは隠されている場所。

![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/ad9d6919-64bb-446f-b219-321e4121cb50)

画面下部にあるように、 `A` を押すと見えるようになる。

yashi commented:
`rm -rf build && west build pico/sample/console-hello -- -DALLOW_DEBUG=1` generates warnigs:

```
warning: The choice symbol USB_DRIVER_LOG_LEVEL_ERR (defined at
subsys/logging/Kconfig.template.log_config:11) was selected (set =y), but no symbol ended up as the
choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_USB_DRIVER_LOG_LEVEL_ERR and/or look up
USB_DRIVER_LOG_LEVEL_ERR in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.


warning: The choice symbol USB_DEVICE_LOG_LEVEL_ERR (defined at
subsys/logging/Kconfig.template.log_config:11) was selected (set =y), but no symbol ended up as the
choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_USB_DEVICE_LOG_LEVEL_ERR and/or look up
USB_DEVICE_LOG_LEVEL_ERR in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.


warning: The choice symbol USB_CDC_ACM_LOG_LEVEL_ERR (defined at
subsys/logging/Kconfig.template.log_config:11) was selected (set =y), but no symbol ended up as the
choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR and/or look up
USB_CDC_ACM_LOG_LEVEL_ERR in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.
```

This also gives you compilation warning:

`rm -rf build && west build pico/sample/console-hello -- -DCONFIG_LOG=y -DCONSOLE=usb`
```
/home/yashi/work/launch2022/rpi-pico/zephyr/subsys/usb/device/class/cdc_acm.c:63:2: warning: #warning "USB_CDC_ACM_LOG_LEVEL forced to LOG_LEVEL_NONE" [-Wcpp]
   63 | #warning "USB_CDC_ACM_LOG_LEVEL forced to LOG_LEVEL_NONE"
      |  ^~~~~~~
```

from `zephyr/subsys/usb/device/class/cdc_acm.c:63`:
```c
#include <zephyr/logging/log.h>
#if DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console), zephyr_cdc_acm_uart) \
	&& defined(CONFIG_USB_CDC_ACM_LOG_LEVEL) \
	&& CONFIG_USB_CDC_ACM_LOG_LEVEL != LOG_LEVEL_NONE
/* Prevent endless recursive logging loop and warn user about it */
#warning "USB_CDC_ACM_LOG_LEVEL forced to LOG_LEVEL_NONE"
#undef CONFIG_USB_CDC_ACM_LOG_LEVEL
#define CONFIG_USB_CDC_ACM_LOG_LEVEL LOG_LEVEL_NONE
#endif
LOG_MODULE_REGISTER(usb_cdc_acm, CONFIG_USB_CDC_ACM_LOG_LEVEL);
```

KantaTamura committed:
pico: sample-console-hello: separate usb logging

通常のデバッグでは不要なusb loggingの項目を
usbを使用した場合のみ有効にするように変更

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura commented:
## `Kconfig`
Zephyr用のコンフィグと追加したloggin用のメニューを追加しています．

## USB_LOGまわり
usb_logの設定をusbとlogを両方有効にしないと無効化するように変更
CDC_ACMでログを発生させないように変更

yashi commented:
`module-str` が `"logging"` だと
![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/f132f685-8f12-4515-b31d-fefae20ba016)
こうなるんだよね。

```
module-str = "Pico Sample App (console-echo)"
```
とかにしない？ 

![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/e6ed5730-76ac-487d-9699-8f1c2be5575a)
シンプルに `"console-echo"` でも良い
![image](https://github.com/spacecubics/scsat1-rpi/assets/10152/0974be86-2dfe-47fa-8ec4-f6c232e481cd)

KantaTamura committed:
pico: sample: delete check DTR process

環境依存で常にDTRが読み込めないことがあるので削除

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura committed:
pico: sample: add build script by twister

```
$ ./zephyr/scripts/twister --testsuite-root scsat1-rpi/pico/sample
```

build-only 2 tasks

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura committed:
.github: workflow: add build test for pico sample program

pico/sample 内のサンプルコードをビルドできるかチェックする

Signed-off-by: KantaTamura <tkanta496@gmail.com>

KantaTamura commented:
```
$ ./zephyr/scripts/twister --testsuite-root scsat1-rpi/pico/sample
```
で，`sample` ディレクトリ内のサンプルプログラムをビルドできるかチェックできます (build-only)

github actionにも追加しています

yashi reviewed (approved):
LGTM
